File: saathi/screen/cards.py
"""On-screen cards for resolving ambiguity by tap or voice — the five
primitives (Choice, Confirm, Read-back, Holding, Entry) and the
controller that shows exactly one of them at a time.

Entry is the fifth, added 2026-09-26 at the user's request: saving a
contact shows the number on a dialpad she can correct by tapping, and
the name, both editable before she says yes. The brief said four; a
read-back with Yes/No made her say the whole number again to fix one
digit. Its answers come in two flavours: edits (`{"number": …}` /
`{"name": …}`) update the card in place with the same id, like a
holding card advancing; a decision (`yes`/`dismiss`) clears it as any
card. Nobody waiting on the card is told about an edit -- the values
travel with the yes.

Why this exists as a module in `screen/` rather than inside each tool
that needs it: a card is the one thing on the screen that asks her
something, and every stream that asks (calling, music, reminders) must
ask the same way — three numbered options at most, spoken as well as
shown, tap or voice both answering, a large way out — or the device has
several dialects of "which one?" and she has to learn each. Streams
import this; they don't draw.

Cards are content and interaction, not status. SPEC.md forbids status
text under the face ("Listening…"); a card is a question she is being
asked, with her options, and the answer is hers to give. Someone will
read "no text on screen" and want to strip this; the distinction is the
whole point of the rule — a label describes the device, a card addresses
her.

Rules the code enforces, not just documents: three options maximum
(`choice()` raises on a fourth — the caller must pick the best three and
say so out loud, which the module can't do for it); every card carries
`spoken`; one card at a time (`show()` replaces); nothing here has a
timer — a card stays until answered, dismissed or replaced; every
non-holding card can be dismissed.

Threading: `show()`/`clear()`/`answer()` are called from the executor
thread (a tool inside `end_turn()`) and from the event-loop thread (the
WebSocket handler on a tap). A lock keeps the current card consistent;
the broadcast goes through the seam `server.py` installs, which hops to
the loop. `ask()` blocks the calling thread until the card is answered.
It must not be called inside `end_turn()`: the turn would sit in
THINKING with the mic closed, so she could only answer by tap, which
breaks "voice and touch both work". Tools show the card and return a
note telling the model what to say; her spoken answer arrives on the
next turn and the tool calls `answer()`. `ask()` is for code that runs
between turns (initiative, a call's own state machine).

Rejected: a component library. None is built for a 75-year-old across a
room; the target sizes, the absence of hover and the one-at-a-time rule
would all be fought rather than given.
"""

from __future__ import annotations

import asyncio
import re
import threading
import uuid
from dataclasses import dataclass, field, replace
from typing import Any, Callable

MAX_OPTIONS = 3
ORDINALS = {1: "One", 2: "Two", 3: "Three"}

KINDS = ("choice", "confirm", "readback", "holding", "entry")


class TooManyOptions(ValueError):
    """More than three. Not truncated here on purpose: the rule is "say
    so and offer the best three", and which three are best is the
    caller's call, as is saying so."""


@dataclass(frozen=True)
class Option:
    n: int  # 1-based; the number she hears and taps
    label: str

    @property
    def spoken(self) -> str:
        return f"{ORDINALS[self.n]}: {self.label}"


@dataclass(frozen=True)
class Card:
    kind: str
    title: str
    spoken: str  # what the engine says when the card appears; never empty
    options: tuple[Option, ...] = ()
    value: str | None = None  # readback: the grouped digits / short value
    progress: float | None = None  # holding: 0..1
    # readback only: the value is a question ("is that right?") answered
    # yes/no by tap or voice, not just acknowledged. See readback().
    confirmable: bool = False
    # entry only: the number as shown (grouped) and the name, both editable.
    number: str | None = None
    name: str | None = None
    id: str = field(default_factory=lambda: uuid.uuid4().hex[:12])

    def as_message(self) -> dict[str, Any]:
        message: dict[str, Any] = {
            "kind": self.kind,
            "id": self.id,
            "title": self.title,
            "spoken": self.spoken,
        }
        if self.kind == "choice":
            message["options"] = [{"n": o.n, "label": o.label} for o in self.options]
        if self.value is not None:
            message["value"] = self.value
        if self.progress is not None:
            message["progress"] = self.progress
        if self.kind == "readback" and self.confirmable:
            message["confirm"] = True
        if self.kind == "entry":
            message["number"] = self.number or ""
            message["name"] = self.name or ""
        return message


# -- builders -------------------------------------------------------------


def choice(title: str, options: list[str] | tuple[str, ...], spoken: str | None = None) -> Card:
    """Two or three numbered options. `spoken` defaults to the title
    followed by each option as "One: …" — exactly what is on screen."""
    labels = [str(label).strip() for label in options if str(label).strip()]
    if len(labels) > MAX_OPTIONS:
        raise TooManyOptions(f"{len(labels)} options; the most a card may carry is {MAX_OPTIONS}")
    if len(labels) < 2:
        raise ValueError("a choice needs two or three options; use confirm() for one")
    opts = tuple(Option(n=i + 1, label=label) for i, label in enumerate(labels))
    if spoken is None:
        spoken = f"{title.strip()} " + " ".join(f"{o.spoken}." for o in opts)
    return Card(kind="choice", title=title.strip(), spoken=spoken.strip(), options=opts)


def confirm(statement: str, spoken: str | None = None) -> Card:
    """One statement, yes or no."""
    statement = statement.strip()
    return Card(kind="confirm", title=statement, spoken=(spoken or statement).strip())


def readback(
    title: str,
    value: str,
    spoken: str | None = None,
    *,
    confirm: bool = False,
    group: bool = True,
) -> Card:
    """A single value shown very large. Digits are grouped for the eye;
    `spoken` defaults to the digits read one at a time with a pause at
    each group, which is how a person reads a number back.

    `confirm=True` makes the read-back a question: Yes/No on screen,
    `{"yes": bool}` accepted by `answer()`. Added at reconciliation for
    Calling, whose read-back before saving a number IS the question --
    the option that lost was a Confirm card carrying the number as its
    statement, which drops the very-large grouped digits a read-back
    exists for. Off by default: a plain read-back is only acknowledged.

    `group=False`: the caller has grouped the value itself and knows its
    structure ("+65 9123 4567" -- country code apart); regrouping in threes
    would give "+659 123 4567". Added at reconciliation for Calling;
    `group_digits` itself keeps normalising, as its own tests require."""
    grouped = group_digits(value) if group else " ".join(str(value).split())
    if spoken is None:
        spoken = f"{title.strip()} {speak_value(grouped)}"
    return Card(
        kind="readback",
        title=title.strip(),
        spoken=spoken.strip(),
        value=grouped,
        confirmable=confirm,
    )


def holding(label: str, progress: float = 0.0, card_id: str | None = None) -> Card:
    """Progress while a button is held. Re-issued with the same id as
    the hold advances so the screen updates in place."""
    progress = max(0.0, min(1.0, float(progress)))
    card = Card(kind="holding", title=label.strip(), spoken=label.strip(), progress=progress)
    if card_id is not None:
        card = Card(**{**card.__dict__, "id": card_id})
    return card


def entry(title: str, *, number: str, name: str, spoken: str | None = None) -> Card:
    """A number on a dialpad and a name, both hers to change before she
    says yes. The number is shown grouped exactly as given (the caller
    knows "+65 9123 4567" has a country code; this module doesn't).

    Name editing is by voice, not an on-screen keyboard: 26 keys that
    fit beside the face would each be far under the 100px floor, and
    spelling by tap is the thing the audience finds hardest. She taps
    "Change name" and says it; the flow behind the card sets it through
    `answer()`."""
    shown = " ".join(str(number).split())
    name = " ".join(str(name).split())
    if spoken is None:
        spoken = (
            f"{title.strip()} {speak_value(shown)}, and the name, {name}. "
            "Say yes to save, or change the number or the name."
        )
    return Card(kind="entry", title=title.strip(), spoken=spoken.strip(), number=shown, name=name)


_RAW_NUMBER = re.compile(r"^\+?\d{0,20}$")


def regroup_number(shown: str, raw: str) -> str:
    """The display for an edited number: the grouping she was shown is
    kept for every leading digit that is unchanged, and digits added
    after that go in fours -- what a tap on the dialpad does, so the
    browser (cards.js, same function) and the server agree without the
    screen knowing what a country code is."""
    kept: list[str] = []
    i = 0
    for ch in shown:
        if ch == " ":
            kept.append(ch)
            continue
        if i < len(raw) and raw[i] == ch:
            kept.append(ch)
            i += 1
        else:
            break
    text = "".join(kept).rstrip()
    for ch in raw[i:]:
        last = text.rsplit(" ", 1)[-1] if text else ""
        if text and len(last.lstrip("+")) >= 4:
            text += " "
        text += ch
    return text


# A phone-number shape: digits with the spacing/punctuation numbers are
# written with. A "." or ":" means a decimal or a time ("37.5",
# "10:30"), which must pass through untouched -- found in review:
# "37.5" was becoming "375".
_DIGITS_ONLY = re.compile(r"^\+?[\d\s\-()]+$")


def group_digits(value: str) -> str:
    """"0412345678" -> "041 234 5678". Anything that isn't a phone-number
    shape (a decimal, a time, a name) passes through untouched. Groups
    of three from the left; a trailing single digit joins the group
    before it (no "… 567 8")."""
    text = str(value).strip()
    if not text or not _DIGITS_ONLY.match(text):
        return text
    prefix = "+" if text.startswith("+") else ""
    digits = re.sub(r"\D", "", text)
    if len(digits) <= 4:
        return prefix + digits
    groups = [digits[i : i + 3] for i in range(0, len(digits), 3)]
    if len(groups[-1]) == 1:
        last = groups.pop()
        groups[-1] += last
    return prefix + " ".join(groups)


def speak_value(grouped: str) -> str:
    """Digits one at a time, a comma between groups; anything else as is."""
    if not _DIGITS_ONLY.match(grouped):
        return grouped
    parts = []
    for group in grouped.split():
        if group.startswith("+"):
            # "plus" is a word, not four letters -- found in review.
            parts.append(("plus " + " ".join(group[1:])).strip())
        else:
            parts.append(" ".join(group))
    return ", ".join(parts)


# -- the controller ------------------------------------------------------


@dataclass(frozen=True)
class Answer:
    card_id: str
    kind: str
    # {"choice": n} | {"yes": bool} | {"dismiss": True}
    # entry: {"yes": True, "number": raw, "name": str} carries her edited values
    answer: dict[str, Any]
    source: str  # "tap" | "voice" | "hold" | "code"
    card: Card

    @property
    def choice(self) -> int | None:
        return self.answer.get("choice")

    @property
    def yes(self) -> bool | None:
        return self.answer.get("yes")

    @property
    def dismissed(self) -> bool:
        return bool(self.answer.get("dismiss"))


Broadcast = Callable[[dict[str, Any]], None]
AnswerCallback = Callable[[Answer], None]


def validate_answer(card: Card, answer: Any) -> dict[str, Any] | None:
    """The one accepted shape per kind, or None. Everything else -- a
    stray key, a fourth option, a string "yes" -- is dropped, never
    guessed at."""
    if not isinstance(answer, dict):
        return None
    if answer.get("dismiss") is True:
        return {"dismiss": True}
    if card.kind == "choice":
        n = answer.get("choice")
        if isinstance(n, bool) or not isinstance(n, int):
            return None
        if 1 <= n <= len(card.options):
            return {"choice": n}
        return None
    if card.kind == "confirm":
        yes = answer.get("yes")
        if isinstance(yes, bool):
            return {"yes": yes}
        return None
    if card.kind == "readback" and card.confirmable:
        yes = answer.get("yes")
        if isinstance(yes, bool):
            return {"yes": yes}
        return None
    if card.kind == "entry":
        return _validate_entry(card, answer)
    return None  # a plain readback and holding only take dismiss


def _raw_number(value: Any) -> str | None:
    if not isinstance(value, str):
        return None
    raw = "".join(value.split())
    return raw if _RAW_NUMBER.match(raw) else None


def _clean_name(value: Any) -> str | None:
    if not isinstance(value, str):
        return None
    name = " ".join(value.split())
    return name or None


def _validate_entry(card: Card, answer: dict[str, Any]) -> dict[str, Any] | None:
    """Yes carries the values to save (hers if given, else the card's
    current ones -- a spoken "yes" doesn't repeat them); no is just no;
    a number and/or name without a decision is an edit."""
    if "yes" in answer:
        yes = answer["yes"]
        if not isinstance(yes, bool):
            return None
        if yes is False:
            return {"yes": False}
        number = _raw_number(answer.get("number", card.number or ""))
        name = _clean_name(answer.get("name", card.name or ""))
        if number is None or name is None:
            return None
        return {"yes": True, "number": number, "name": name}
    edit: dict[str, Any] = {}
    if "number" in answer:
        number = _raw_number(answer["number"])
        if number is None:
            return None
        edit["number"] = number
    if "name" in answer:
        name = _clean_name(answer["name"])
        if name is None:
            return None
        edit["name"] = name
    return edit or None


def is_edit(answer: dict[str, Any]) -> bool:
    """An entry card's number and/or name without a decision."""
    return bool(answer) and set(answer) <= {"number", "name"}


class CardController:
    """Holds the one current card and routes its answer.

    `answer()` is the single entry point for both ways of answering:
    the WebSocket handler calls it on a tap (`source="tap"`), a tool
    calls it when she says "the second one" (`source="voice"`). Either
    clears the card and wakes every `on_answer` callback and any
    `ask()` waiting on it."""

    def __init__(self, broadcast: Broadcast | None = None) -> None:
        self._broadcast = broadcast
        self._lock = threading.Lock()
        self._current: Card | None = None
        self._callbacks: list[AnswerCallback] = []
        self._waiters: dict[str, tuple[threading.Event, list[Answer]]] = {}

    def set_broadcast(self, broadcast: Broadcast | None) -> None:
        self._broadcast = broadcast

    @property
    def current(self) -> Card | None:
        return self._current

    def _emit(self, card: Card | None) -> None:
        if self._broadcast is None:
            return
        self._broadcast({"type": "card", "card": card.as_message() if card else None})

    def show(self, card: Card) -> str:
        """Show `card`, replacing whatever was up. One at a time, never
        stacked: a replaced card's `ask()` is released with a dismiss
        so nothing waits on a question she can no longer see."""
        if card.kind not in KINDS:
            raise ValueError(f"unknown card kind {card.kind!r}")
        if not card.spoken:
            raise ValueError("every card must have something to say")
        # The broadcast happens under the lock so the screen is told
        # in the same order `_current` changed -- two shows racing from
        # the executor and loop threads must not leave the screen on a
        # card whose taps are then rejected as stale (found in review).
        # Callbacks run outside it: one may show() a card of its own.
        with self._lock:
            previous = self._current
            self._current = card
            replaced = None
            if previous is not None and previous.id != card.id:
                replaced = self._release(previous, {"dismiss": True}, "code")
            self._emit(card)
        if replaced is not None:
            self._notify(replaced)
        return card.id

    def clear(self) -> None:
        with self._lock:
            previous = self._current
            self._current = None
            released = self._release(previous, {"dismiss": True}, "code") if previous else None
            self._emit(None)
        if released is not None:
            self._notify(released)

    def answer(self, card_id: str, answer: Any, *, source: str = "tap") -> bool:
        """True if `card_id` is the current card and `answer` is a valid
        shape for it; the card is then cleared and everyone waiting is
        told. False (and nothing happens) for a stale id -- a tap on a
        card that was already replaced -- or a malformed answer."""
        with self._lock:
            card = self._current
            if card is None or card.id != card_id:
                return False
            valid = validate_answer(card, answer)
            if valid is None:
                return False
            if is_edit(valid):
                # An entry card edited, not decided: same id, new values,
                # still up. Nobody waiting is told; the yes carries them.
                updated = replace(
                    card,
                    number=regroup_number(card.number or "", valid["number"])
                    if "number" in valid
                    else card.number,
                    name=valid.get("name", card.name),
                )
                self._current = updated
                self._emit(updated)
                return True
            self._current = None
            result = self._release(card, valid, source)
            self._emit(None)
        self._notify(result)
        return True

    def on_answer(self, callback: AnswerCallback) -> Callable[[], None]:
        self._callbacks.append(callback)

        def unsubscribe() -> None:
            if callback in self._callbacks:
                self._callbacks.remove(callback)

        return unsubscribe

    def _release(self, card: Card, answer: dict[str, Any], source: str) -> Answer:
        """Under the lock: build the Answer and wake an ask() waiter."""
        result = Answer(card_id=card.id, kind=card.kind, answer=answer, source=source, card=card)
        waiter = self._waiters.pop(card.id, None)
        if waiter is not None:
            event, slot = waiter
            slot.append(result)
            event.set()
        return result

    def _notify(self, result: Answer) -> None:
        for callback in list(self._callbacks):
            callback(result)

    def ask(self, card: Card, timeout: float | None = None) -> Answer | None:
        """Show `card` and block until it is answered, dismissed or
        replaced. No timeout by default -- nothing on a card times out.
        Not for use inside `end_turn()` (see the module docstring).
        Returns None only if a `timeout` was given and passed; the card
        is then cleared."""
        event = threading.Event()
        slot: list[Answer] = []
        with self._lock:
            self._waiters[card.id] = (event, slot)
        self.show(card)
        if not event.wait(timeout):
            with self._lock:
                self._waiters.pop(card.id, None)
                still_up = self._current is not None and self._current.id == card.id
                if still_up:
                    self._current = None
            if still_up:
                self._emit(None)
            return None
        return slot[0]

    async def ask_async(self, card: Card, timeout: float | None = None) -> Answer | None:
        """`ask()` for code on the event loop (initiative, a call's own
        loop-side state machine)."""
        loop = asyncio.get_running_loop()
        return await loop.run_in_executor(None, self.ask, card, timeout)


# -- hold-to-confirm ----------------------------------------------------------

HoldComplete = Callable[[], None]


class HoldController:
    """The spacebar as a hold-to-confirm button (a call's hang-up).

    While a handler is set, `server.py` routes press/release here
    instead of to `core.py`: a short press does nothing at all -- no
    turn, no capture -- and a press held for `seconds` fires
    `on_complete()` exactly once. The Holding card shows the progress
    of the hold so she can see it is doing something and let go if she
    didn't mean it. The timer lives in the server (it owns the loop);
    this object only knows how far along the hold is and whether it
    has fired.

    Why the press is taken away from the state machine entirely rather
    than passed through: during a call the mic is already open to the
    other person, and a press starting a turn would talk over the call.
    The stream that sets the handler is saying "for now, the button
    means this"."""

    def __init__(self, cards: CardController) -> None:
        self._cards = cards
        self._handler: HoldComplete | None = None
        self.seconds = 2.0
        self.label = "Keep holding"
        self._card_id: str | None = None
        self._fired = False

    def set_handler(
        self, on_complete: HoldComplete, *, seconds: float = 2.0, label: str = "Keep holding"
    ) -> None:
        if seconds <= 0:
            raise ValueError("seconds must be positive")
        self._handler = on_complete
        self.seconds = float(seconds)
        self.label = label.strip() or "Keep holding"

    def clear(self) -> None:
        self._handler = None
        self.abandon()

    @property
    def active(self) -> bool:
        return self._handler is not None

    @property
    def holding(self) -> bool:
        return self._card_id is not None

    def begin(self) -> None:
        """The press. Shows the Holding card at 0."""
        self._fired = False
        card = holding(self.label, 0.0)
        self._card_id = card.id
        self._cards.show(card)

    def tick(self, elapsed: float) -> bool:
        """Called by the server's timer. Updates the card; at `seconds`
        fires the handler once and clears the card. Returns True once
        complete."""
        if self._card_id is None or self._fired:
            return self._fired
        progress = min(1.0, elapsed / self.seconds)
        if progress >= 1.0:
            self._fired = True
            handler = self._handler
            card_id = self._card_id
            self._card_id = None
            current = self._cards.current
            if current is not None and current.id == card_id:
                self._cards.clear()
            if handler is not None:
                handler()
            return True
        self._cards.show(holding(self.label, progress, card_id=self._card_id))
        return False

    def abandon(self) -> None:
        """The release before `seconds`: the card goes, nothing fires."""
        card_id = self._card_id
        self._card_id = None
        if card_id is None:
            return
        current = self._cards.current
        if current is not None and current.id == card_id:
            self._cards.clear()

File: saathi/screen/static/js/cards.js
// Cards: the five primitives for resolving ambiguity by tap or voice —
// Choice, Confirm, Read-back, Holding, Entry — drawn beside the face,
// never in its place. Entry (2026-09-26, the user's ask) is the number
// on a dialpad and the name, both hers to correct before Save: a digit
// key edits the number here at once and sends the edit; the server
// re-issues the card with the same id and the new values (a voice
// edit arrives the same way). The name is changed by voice -- an
// on-screen keyboard's keys would be a third the size of these. The browser counterpart of screen/cards.py; the module the
// other streams' screens go through. Nothing here decides what to ask:
// a card arrives as a `card` message from the server (a tool showed it)
// and a tap goes back as `card_answer`. A spoken answer never touches
// this file — it goes through a tool to the same CardController.
//
// Plain HTML and CSS on purpose. Every component library is built for a
// sighted adult with a mouse at arm's length; this is a 75-year-old
// across a room. The rules this file keeps (each load-bearing, from the
// brief): three options at most, numbered so "the second one" works;
// touch targets at least 100px tall; nothing smaller than 32px; AAA
// contrast (tokens in style.css, checked by tests/test_cards_style.py);
// no hover states; nothing times out — there is no setTimeout in this
// file, and tests grep for one; one card at a time — a new card
// replaces the old, never stacks; a large way out on every card that
// can be answered ("Never mind" / "Okay"), never a small ✕; tap only —
// no sliders, drag or long-press. Holding is the one card with no
// button: letting go of the spacebar is its way out.
//
// Cards are content and interaction, not status. "Listening…" describes
// the device and is forbidden (SPEC.md, CLAUDE.md); "Which one?" with
// three numbered names addresses her and waits for her answer.

const KINDS = new Set(["choice", "confirm", "readback", "holding", "entry"]);

const KEYS = ["1", "2", "3", "4", "5", "6", "7", "8", "9", "⌫", "0", "+"];

// The raw digits (and a leading +) of a shown, grouped number.
function rawOf(shown) {
  return String(shown).replace(/\s+/g, "");
}

// The display for an edited number: keep the grouping she was shown
// for every leading digit that is unchanged; digits added after that
// go in fours. Same function as regroup_number in screen/cards.py, so
// a tap here and the server's re-issue agree.
function regroup(shown, raw) {
  let kept = "";
  let i = 0;
  for (const ch of shown) {
    if (ch === " ") {
      kept += ch;
      continue;
    }
    if (i < raw.length && raw[i] === ch) {
      kept += ch;
      i += 1;
    } else {
      break;
    }
  }
  let text = kept.replace(/\s+$/, "");
  for (const ch of raw.slice(i)) {
    const last = text ? text.split(" ").pop() : "";
    if (text && last.replace(/^\+/, "").length >= 4) text += " ";
    text += ch;
  }
  return text;
}

const DEMO = {
  "cards-choice": {
    kind: "choice",
    id: "demo-choice",
    title: "Which Priya?",
    spoken: "Which Priya? One: Priya, your daughter. Two: Priya from church. Three: Priya Menon.",
    options: [
      { n: 1, label: "Priya, your daughter" },
      { n: 2, label: "Priya from church" },
      { n: 3, label: "Priya Menon" },
    ],
  },
  "cards-confirm": {
    kind: "confirm",
    id: "demo-confirm",
    title: "Call Priya, your daughter?",
    spoken: "Call Priya, your daughter?",
  },
  "cards-readback": {
    kind: "readback",
    id: "demo-readback",
    title: "Priya's number",
    spoken: "Priya's number: zero four one, two three four, five six seven eight",
    value: "041 234 5678",
  },
  "cards-holding": {
    kind: "holding",
    id: "demo-holding",
    title: "Keep holding to hang up",
    spoken: "Keep holding to hang up",
    progress: 0.6,
  },
  "cards-entry": {
    kind: "entry",
    id: "demo-entry",
    title: "Priya's number",
    spoken:
      "Priya's number plus six five, nine one two three, four five six seven, and the name, " +
      "Priya. Say yes to save, or change the number or the name.",
    number: "+65 9123 4567",
    name: "Priya",
  },
};

function button(className, text, onTap) {
  const el = document.createElement("button");
  el.type = "button";
  el.className = className;
  el.textContent = text;
  el.addEventListener("click", onTap);
  return el;
}

export function createCards(send, options = {}) {
  const demo = options.demo || null;
  let current = null; // the card message currently shown, or null

  const root = document.createElement("div");
  root.id = "card-panel";
  root.hidden = true;
  document.body.appendChild(root);

  function answer(cardId, payload) {
    send({ type: "card_answer", id: cardId, answer: payload });
  }

  function renderChoice(card, into) {
    const list = document.createElement("div");
    list.className = "card__options";
    for (const option of card.options.slice(0, 3)) {
      const el = button("card__option", "", () => answer(card.id, { choice: option.n }));
      const number = document.createElement("span");
      number.className = "card__number";
      number.textContent = String(option.n);
      const label = document.createElement("span");
      label.className = "card__label";
      label.textContent = option.label;
      el.replaceChildren(number, label);
      list.appendChild(el);
    }
    into.appendChild(list);
    into.appendChild(button("card__dismiss", "Never mind", () => answer(card.id, { dismiss: true })));
  }

  function renderConfirm(card, into) {
    const row = document.createElement("div");
    row.className = "card__options card__options--row";
    row.appendChild(button("card__option card__option--yes", "Yes", () => answer(card.id, { yes: true })));
    row.appendChild(button("card__option card__option--no", "No", () => answer(card.id, { yes: false })));
    into.appendChild(row);
    into.appendChild(button("card__dismiss", "Never mind", () => answer(card.id, { dismiss: true })));
  }

  function renderReadback(card, into) {
    const value = document.createElement("div");
    value.className = "card__value";
    value.textContent = card.value == null ? "" : String(card.value);
    into.appendChild(value);
    if (card.confirm) {
      // A read-back that is a question ("is that right?"): Yes / No,
      // and still a way out. Same buttons and sizes as a Confirm card.
      const row = document.createElement("div");
      row.className = "card__options card__options--row";
      row.appendChild(button("card__option card__option--yes", "Yes", () => answer(card.id, { yes: true })));
      row.appendChild(button("card__option card__option--no", "No", () => answer(card.id, { yes: false })));
      into.appendChild(row);
      into.appendChild(button("card__dismiss", "Never mind", () => answer(card.id, { dismiss: true })));
      return;
    }
    into.appendChild(button("card__dismiss", "Okay", () => answer(card.id, { dismiss: true })));
  }

  function renderEntry(card, into) {
    let shown = String(card.number);
    let name = String(card.name);

    const number = document.createElement("div");
    number.className = "card__value card__value--entry";
    number.textContent = shown;
    into.appendChild(number);

    const body = document.createElement("div");
    body.className = "card__entry";

    const pad = document.createElement("div");
    pad.className = "card__dialpad";
    const edit = (raw) => {
      shown = regroup(shown, raw);
      number.textContent = shown;
      answer(card.id, { number: raw });
    };
    for (const key of KEYS) {
      pad.appendChild(
        button("card__key", key, () => {
          const raw = rawOf(shown);
          if (key === "⌫") {
            if (raw) edit(raw.slice(0, -1));
          } else if (key === "+") {
            if (!raw) edit("+");
          } else {
            edit(raw + key);
          }
        })
      );
    }
    body.appendChild(pad);

    const side = document.createElement("div");
    side.className = "card__entry-side";
    const nameRow = document.createElement("div");
    nameRow.className = "card__name";
    const nameLabel = document.createElement("span");
    nameLabel.className = "card__name-label";
    nameLabel.textContent = "Name: ";
    const nameText = document.createElement("span");
    nameText.className = "card__name-text";
    nameText.textContent = name;
    nameRow.replaceChildren(nameLabel, nameText);
    side.appendChild(nameRow);
    const hint = document.createElement("div");
    hint.className = "card__hint";
    hint.textContent = "Say the new name.";
    hint.hidden = true;
    side.appendChild(button("card__option card__option--rename", "Change name", () => {
      hint.hidden = false;
    }));
    side.appendChild(hint);
    side.appendChild(
      button("card__option card__option--yes card__option--save", "Save", () =>
        answer(card.id, { yes: true, number: rawOf(shown), name })
      )
    );
    side.appendChild(
      button("card__option card__option--no", "Try again", () => answer(card.id, { yes: false }))
    );
    side.appendChild(button("card__dismiss", "Never mind", () => answer(card.id, { dismiss: true })));
    body.appendChild(side);
    into.appendChild(body);

    // The server re-issued the same card with new values (a voice
    // edit, or the echo of a tap): update in place, keep the dialpad.
    into.update = (next) => {
      shown = String(next.number);
      name = String(next.name);
      number.textContent = shown;
      nameText.textContent = name;
      hint.hidden = true;
    };
  }

  function renderHolding(card, into) {
    const track = document.createElement("div");
    track.className = "card__progress";
    const fill = document.createElement("div");
    fill.className = "card__progress-fill";
    const progress = Math.max(0, Math.min(1, Number(card.progress) || 0));
    fill.style.width = `${Math.round(progress * 100)}%`;
    track.appendChild(fill);
    into.appendChild(track);
  }

  function render() {
    root.replaceChildren();
    if (current === null) {
      root.hidden = true;
      document.body.classList.remove("card--open");
      window.dispatchEvent(new Event("resize"));
      return;
    }
    const card = document.createElement("div");
    card.className = `card card--${current.kind}`;
    card.dataset.cardId = current.id;
    const title = document.createElement("div");
    title.className = "card__title";
    title.textContent = current.title || "";
    card.appendChild(title);
    switch (current.kind) {
      case "choice":
        renderChoice(current, card);
        break;
      case "confirm":
        renderConfirm(current, card);
        break;
      case "readback":
        renderReadback(current, card);
        break;
      case "holding":
        renderHolding(current, card);
        break;
      case "entry":
        renderEntry(current, card);
        break;
      default:
        break;
    }
    root.replaceChildren(card);
    root.hidden = false;
    document.body.classList.add("card--open");
    // The face's container may have changed size (eyes-face.js only
    // reflows on a window resize).
    window.dispatchEvent(new Event("resize"));
  }

  function onCard(message) {
    const card = message.card;
    if (card === null || card === undefined) {
      current = null;
      render();
      return;
    }
    if (typeof card !== "object" || !KINDS.has(card.kind) || typeof card.id !== "string") {
      return; // malformed: dropped, the current card stays
    }
    if (card.kind === "choice" && !Array.isArray(card.options)) return;
    if (card.kind === "entry" && (typeof card.number !== "string" || typeof card.name !== "string")) {
      return;
    }
    if (current !== null && current.id === card.id && current.kind === "entry" && card.kind === "entry") {
      // The same entry card with edited values: update in place, so
      // the dialpad she is tapping on isn't rebuilt under her finger.
      current = card;
      const node = root.querySelector(".card");
      if (node && typeof node.update === "function") {
        node.update(card);
        return;
      }
    }
    if (current !== null && current.id === card.id && current.kind === "holding") {
      // Same hold advancing: update the bar in place rather than
      // rebuilding the card twenty times a hold.
      current = card;
      const fill = root.querySelector(".card__progress-fill");
      if (fill) {
        const progress = Math.max(0, Math.min(1, Number(card.progress) || 0));
        fill.style.width = `${Math.round(progress * 100)}%`;
        return;
      }
    }
    current = card; // one at a time: whatever was up is replaced
    render();
  }

  if (demo && DEMO[demo]) {
    onCard({ type: "card", card: DEMO[demo] });
  }

  return {
    onMessage(message) {
      if (message.type === "card") onCard(message);
    },
    // For tests only.
    _current() {
      return current;
    },
  };
}
